### PR TITLE
MLD customization: fill weights in CellStorage

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

PR partially addresses #3350: filling weights in CellStorage.

## Tasklist
 - [x] add a simple one-to-many Dijkstra cell customizer
 - [x] add `osrm-customize` tool
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
Required PRs: #3724, #3730

### CODE REVIEW

File: include/customizer/cell_customizer.hpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                   osrm::util::LevelID level,
+                   osrm::util::CellID id)
+    {
+        auto cell = cells.GetCell(level, id);
+        auto destinations = cell.GetDestinationNodes();
+
+        // for each source do forward search
+        for (auto source : cell.GetSourceNodes())
+        {
+            std::unordered_set<NodeID> destinations_set(destinations.begin(), destinations.end());
```
Comment (TheMarex) [inline include/customizer/cell_customizer.hpp]: Destination nodes should be unique, why do we need the unordered_set here?

Comment (oxidase) [inline include/customizer/cell_customizer.hpp]: Here it is just a data structure with fast erase operation to stop search if all destination nodes are reached.

**Comment by oxidase** [comment]
Rebased onto `mld/facade` branch. Added only unit tests,  proper routing testing would require an MLD routing plugin.

File: include/customizer/cell_customizer.hpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+#include <boost/thread/tss.hpp>
+#include <unordered_set>
+
+namespace osrm
+{
+namespace customizer
+{
+
+class CellCustomizer
```
Comment (TheMarex) [inline include/customizer/cell_customizer.hpp]: This class does not maintain any state right now. Do you think we could just split it into free functions?

File: include/partition/cell_storage.hpp [hunk 3]
```diff
@@ -117,33 +123,27 @@ template <bool UseShareMemory> class CellStorageImpl
             std::size_t stride;
         };
 
-        std::size_t GetRow(NodeID node) const
```
Comment (TheMarex) [inline include/partition/cell_storage.hpp]: What is the idea here behind inlining these functions below and removing the interface?

File: unit_tests/util/cell_customization.cpp [hunk 4]
```diff
@@ -0,0 +1,277 @@
+#include <boost/numeric/conversion/cast.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include "customizer/cell_customizer.hpp"
+#include "partition/cell_storage.hpp"
+#include "partition/multi_level_partition.hpp"
+#include "util/static_graph.hpp"
+
+#define REQUIRE_SIZE_RANGE(range, ref) BOOST_REQUIRE_EQUAL(range.size(), ref)
+#define CHECK_EQUAL_RANGE(range, ...)                                                              \
```
Comment (TheMarex) [inline unit_tests/util/cell_customization.cpp]: This is cool, could we move this into a common header? We could reuse this for the `CellStorage` tests.

File: include/customizer/cell_storage.hpp [hunk 5]
```diff
@@ -186,24 +186,24 @@ template <bool UseShareMemory> class CellStorageImpl
     CellStorageImpl() {}
 
     template <typename GraphT, typename = std::enable_if<!UseShareMemory>>
-    CellStorageImpl(const MultiLevelPartition &partition, const GraphT &base_graph)
+    CellStorageImpl(const partition::MultiLevelPartition &partition, const GraphT &base_graph)
     {
         // pre-allocate storge for CellData so we can have random access to it by cell id
         unsigned number_of_cells = 0;
-        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        for (partition::LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
```
Comment (TheMarex) [inline include/customizer/cell_storage.hpp]: Should we move `LevelID` and `CellID` to the global OSRM namespace?

File: include/partition/multi_level_partition.hpp [hunk 6]
```diff
@@ -33,14 +33,13 @@ using MultiLevelPartitionView = detail::MultiLevelPartitionImpl<true>;
...
 {
 template <bool UseShareMemory>
+void read(const boost::filesystem::path &file,
+          detail::MultiLevelPartitionImpl<UseShareMemory> &mlp);
+template <bool UseShareMemory>
 void write(const boost::filesystem::path &file,
            const detail::MultiLevelPartitionImpl<UseShareMemory> &mlp);
 }
 
-namespace detail
```
Comment (TheMarex) [inline include/partition/multi_level_partition.hpp]: Good catch, thank you.

File: src/partition/partitioner.cpp [hunk 7]
```diff
@@ -174,14 +174,8 @@ int Partitioner::Run(const PartitionConfig &config)
     TIMER_STOP(packed_mlp);
     util::Log() << "MultiLevelPartition constructed in " << TIMER_SEC(packed_mlp) << " seconds";
 
-    TIMER_START(cell_storage);
```
Comment (TheMarex) [inline src/partition/partitioner.cpp]: Eventually we might want to move this back as generating this data structure takes some time (for large datasets) and we don't want it in the hot-path of `osrm-customize` runs.

Thinking long-term making `osrm-customize` eventually digest `.osrm.cells` will will enable only updating cells selectively and using an earlier customization for the rest.

Comment (oxidase) [inline src/partition/partitioner.cpp]: Sure, just forgot about heavy constructor. I will add a reader function for `CellStorage`

File: include/partition/cell_storage.hpp (hunk 3 above)
Comment (oxidase) [inline include/partition/cell_storage.hpp]: `GetOutWeight` and `GetInWeight` will return an empty range if `node`  is not a source or destination node. Row or column value in this case should be an invalid value and introduce an undefined behavior, so i moved both `find` calls and iterator checks to `GetOutWeight` and `GetInWeight`

**Pull request merged**